Matlab/Matlab.sublime-syntax

```diff
@@ -48,477 +76,792 @@ contexts:
-      captures:
-        1: constant.numeric.base.matlab
-        2: constant.numeric.value.matlab
-        3: constant.numeric.suffix.matlab
-    - match: '\b(0[bB])([01]+)(u8|u16|u32|u64|s8|s16|s32|s64)?\b'
-      scope: meta.number.integer.binary.matlab
-      captures:
-        1: constant.numeric.base.matlab
-        2: constant.numeric.value.matlab
-        3: constant.numeric.suffix.matlab
-    - match: '((?:\d*(\.))?\d+(?:[Ee][-+]?\d+)?)(i|j)\b'
-      scope: meta.number.imaginary.decimal.matlab
-      captures:
-        1: constant.numeric.value.matlab
-        2: punctuation.separator.decimal.matlab
-        3: constant.numeric.suffix.matlab
-    - match: '(?:\d*(\.))?\d+(?:[Ee][-+]?\d+)?\b'
-      scope: meta.number.float.decimal.matlab constant.numeric.value.matlab
-      captures:
-        1: punctuation.separator.decimal.matlab
-  operators:
-    - match: \s*(==|~=|~|>|>=|<|<=|=|&|&&|:|\||\|\||\+|-|\*|\.\*|/|\./|\\|\.\\|\^|\.\^)\s*
-      comment: Operator symbols
-      scope: keyword.operator.symbols.matlab
-  parens:
-    - match: \(
-      scope: punctuation.section.parens.begin.matlab
+        - include: keywords
+        - include: expressions
+
+  for:
+    - match: \bfor\b
+      scope: keyword.control.loop.for.matlab
       push:
-        - meta_scope: meta.parens.matlab
-        - match: \)
-          scope: punctuation.section.parens.end.matlab
-          # pop: true
-          set: transpose_post_parens
-        - include: allofem
-        - include: end_in_parens
-  special_characters:
-    - match: '((\%([\+\-0]?\d{0,3}(\.\d{1,3})?)(c|d|e|E|f|g|G|s|((b|t)?(o|u|x|X))))|\%\%|\\(b|f|n|r|t|\\))'
-      comment: Operator symbols
-      scope: constant.character.escape.matlab
-  string:
-    - match: ''''
-      scope: punctuation.definition.string.begin.matlab
+        - meta_scope: meta.block.for.matlab
+        - match: \bend\b
+          scope: keyword.control.end.for.matlab
+          pop: true
+        - include: keywords
+        - include: expressions
+
+  parfor:
+    - match: \bparfor\b
+      scope: keyword.control.loop.parfor.matlab
       push:
-        - meta_scope: string.quoted.single.matlab
-        - match: '''(?=(\]|\)|\}|=|~|<|>|&|\||-|\+|\*|\.|\^|\||\s|;|:|,))'
-          scope: punctuation.definition.string.end.matlab
+        - meta_scope: meta.block.parfor.matlab
+        - match: \bend\b
+          scope: keyword.control.end.parfor.matlab
           pop: true
-        - include: escaped_quote
-        - include: unescaped_quote
-        - include: special_characters
-    - match: '"'
-      scope: punctuation.definition.string.begin.matlab
+        - include: keywords
+        - include: expressions
+
+  while:
+    - match: \bwhile\b
+      scope: keyword.control.loop.while.matlab
       push:
-        - meta_scope: string.quoted.double.matlab
-        - match: '"'
-          scope: punctuation.definition.string.end.matlab
+        - meta_scope: meta.block.while.matlab
```

Comment:
These kinds of use cases of sub-scoping `meta.block` are quite interesting. While this style is found in older syntaxes (e.g.AppleScript) most recently modified syntaxes seem to just use `meta.block.<syntax>` without sub-scoping. 

They rather use `meta.while meta.block`, maybe because the loop condition `while (condition)` doesn't really belong to the block, but may rather be interpreted as some kind of argument for the while keyword. The block is only one of several parts of such a statement. While it is easy to apply for syntaxes which use `{` `}` to denote blocks, I am a bit uncertain about the correct style with regards to syntaxes like matlab or ruby, ... which use keywords only.

I am also not very certain about `meta.while meta.block` is the best solution. I use it in the rewritten Java, but it quite quickly fills the scope stack because of each level of statement adding 2 meta scopes.

Reply:
I assume the restriction for `meta.block` to `{ ... }` was due to it beeing one of the scope recommendations in https://www.sublimetext.com/docs/scope_naming.html#meta.

Some languages use an additional `begin` keyword, e.g.
```julia
while condition begin
   ...
end
```
so it would be analogous here to have `meta.block` from (and including) `begin` to `end`. But I don't really think there is a big advantage to have separate `meta.while meta.block` scopes (at least in Matlab which doesn't use a `begin` keyword). And I when I say "block" here, I would probably mean the whole statement, including `while` keyword and condition. But it's probably a matter of interpretation.

Anyway, I would suggest to keep a single `meta.block.while` here.